src/Raven.Server/Documents/Document.cs

```diff
@@ -124,15 +125,25 @@ public Document Clone(JsonOperationContext context)
             var newData = Data.Clone(context);
             return CloneWith(context, newData);
         }
+        
+        public virtual TDocument Clone<TDocument>(JsonOperationContext context)
+            where TDocument : Document, new()
+        {
+            var newData = Data.Clone(context);
+            return CloneWith<TDocument>(context, newData);
+        }
 
-        public Document CloneWith(JsonOperationContext context, BlittableJsonReaderObject newData)
+        public Document CloneWith(JsonOperationContext context, BlittableJsonReaderObject newData) => CloneWith<Document>(context, newData);
+        
+        public virtual TDocument CloneWith<TDocument>(JsonOperationContext context, BlittableJsonReaderObject newData)
+            where TDocument : Document, new()
         {
             var newId = context.GetLazyString(Id);
             var newLowerId = context.GetLazyString(LowerId);
-            return new Document
+            return new TDocument()
             {
                 Etag = Etag,
-                StorageId = StorageId,
+                StorageId = Voron.Global.Constants.Compression.NonReturnableStorageId, // this is basically in-memory copy, so we should care about storage...
```

Comment:
No, the decompressed buffer is owned by the first instance of document pushed into the LRU cache. All copies are storage independent (because it's in-memory clone).